### ISSUE TITLE
Support accessible output

### ISSUE THREAD

**Issue by JadedBlueEyes** [typst/typst#133]
This sort of links to #114 and #129 - Typst should support accessible output formatting - for example, tagged PDF. Latex does not support this, and so people and organisations that want to produce accessible documents are forced to manually mark up documents in Adobe Acrobat. 

Typst outputting accessible documents would be a huge win both for the disabled community and organisations and individuals that produce documents. I think it would make people a lot more likely to use and suggest the project as well.

For some more context, the LaTeX Project Team have been working on trying to get something like this in LaTeX for a while - https://www.pdfa.org/presentation/tagged-and-accessible-pdf-with-latex/

**Comment by hcsch** [comment]
https://www.pdfa.org/resource/deriving-html-from-pdf/ is a best practices guide for tagged PDF syntax (also covers some general information/guidances around the format), and with regards to #114 and this issue https://www.pdfa.org/resource/deriving-html-from-pdf/ is perhaps also an interesting document.

**Comment by jasonjgw** [comment]
Based on my own experience, combined with various discussions on the blindmath mailing list and elsewhere, there is general agreement that HTML output is far preferable to screen reader users than PDF of any kind, whether tagged or not. That is, for users who are blind, producing high-quality HTML output would be the best path to take (preferably with MathML included).

Most PDF reading applications don't even interpret the tags, where present, and the reading experience doesn't match what an HTML browser can do, on multiple platforms and with multiple screen reading technologies, desktop and mobile alike.

Based on the reference documentation for Typst, at the moment there are no function parameters for specifying a short description of an image - what would appear in an HTML ALT attribute. Also, the digital publishing accessibility community has developed best practices for including extended descriptions in documents. These descriptions can include entire document structures (e.g., tables, paragraphs, etc.). At the moment, the favoured technique is to use the HTML 5 `summary`/`details` elements, with some WAI-ARIA markup. There are other approaches as well. Many figures in technical and scientific documents would probably require, or at least benefit from extended descriptions, which authors should be able to include in the source markup.

Another accessibility requirement (currently a problem for LaTeX) is to distinguish between header and body cells in tables.

I'm sure there will be further issues to consider. I would be interested in testing the HTML export functionality when it is integrated into the compiler, and in contributing comments. At the moment, I use LaTeX for much of my writing, from which I generate HTML for review with a screen rader (i.e., interactive braille output and text to speech).

**Comment by u-fischer** [comment]
>  for example, tagged PDF. Latex does not support this

This is not true. LaTeX supports tagged PDF since quite some time. The documentation of tagpdf and the texlive vpat e.g. has been tagged with LaTeX, and there exist also some non-public implementations (e.g. from Ross Moore and from a publisher in Germany).  The goal of the LaTeX Project is not to make tagging *possible*, but to make it automatic, free and easy to use so that an average user can handle it (and since the latest latex-dev release this already works quite well for simple documents, and it will get better until the summer).

**Comment by SeSodesa** [comment]
The lack of [PDF/UA] support is also a blocker for producing public documents, such as study materials for consumption in the European Union. This is due to the newish EU accessibility directive ([link]).

[PDF/UA]: https://en.wikipedia.org/wiki/PDF/UA
[link]: https://eur-lex.europa.eu/legal-content/EN/TXT/HTML/?uri=CELEX:52015PC0615

**Comment by laurmaedje** [comment]
Just because there was no "official" response in this thread before: We definitely plan to make Typst's output accessible in the future, we just haven't gotten to it yet. It needs some work on compiler internals, but overall I believe we're in a good position because there is quite a lot of semantic information available in Typst documents. The necessary reworks are also partially shared with those for HTML export, which is good.

**Comment by acxz** [comment]
> > for example, tagged PDF. Latex does not support this
> 
> This is not true. LaTeX supports tagged PDF since quite some time. The documentation of tagpdf and the texlive vpat e.g. has been tagged with LaTeX, and there exist also some non-public implementations (e.g. from Ross Moore and from a publisher in Germany). The goal of the LaTeX Project is not to make tagging _possible_, but to make it automatic, free and easy to use so that an average user can handle it (and since the latest latex-dev release this already works quite well for simple documents, and it will get better until the summer).

Just a follow up to this and to add an additional resource:
LaTeX examples on conforming to various PDF Accessibility (including PDF/UA-2)
https://github.com/latex3/tagging-project/discussions/72

**Comment by ErroneousBosch** [comment]
My day job is as a web developer and accessibility professional for a major academic research institution, and in evaluating LaTeX tagging for our faculty I found that it is not in a usable state either using the tagpdf package or using the core tagging via TeXLive. One of the major issues is decision to program to the PDF/UA-2 standard, which isn't implemented by any currently available reader software even four years after its release. 

As I point out in the thread @acxz links to, compliance to an unused/unimplemented standard means that in practical terms the PDFs it generates aren't accessible, and aren't going to be in any useful timeframe. It is so bad in fact that we are recommending that our faculty no longer use it for class materials at all, and we are trying to figure out a solution for research papers. I would urge the developers here to not follow LaTeX's decision path. PDF/A-2 may end up a great solution if it is ever widely adopted, but for right now it isn't useful to set as a goal.

**Comment by u-fischer** [comment]
@ErroneousBosch as we told you in the thread: yes we push and promote PDF 2.0 and the PDF/UA-2 standard as you can't get sufficiently accessible math with the older standard. And we are not doing this alone: the documents were created at the request and in cooperation with the PDF association https://pdfa.org/wtpdf/. 

But we did not drop support for the older standard in the process. If you prefer to create PDF/UA-1 the only thing you must do is to change 

~~~~
\DocumentMetadata{
   pdfversion=2.0,
   pdfstandard=ua-2,
  ....
~~~~

into 

~~~~
\DocumentMetadata{
   pdfversion=1.7,
   pdfstandard=ua-1,
  ....
~~~~

Or you can stay with PDF 2.0 and test what happens. If possible we implement compability options and for example added a fallback to 1.7 names to that the files can be used also with tools that do not properly support PDF 2.0. If you know of an example where this is not the case we would be happy to hear about it so that we can look how to improve compability with older tools.

**Comment by ErroneousBosch** [comment]
@u-fischer I did test the PDF 1.7 generation, and as you say, it does not produce accessible math. For our use cases and those of many other LaTeX users in academia, PDF 2.0 isn't useful. We cannot implement it and then wait for it to be widely accepted; the institutional and legal requirements we have to comply with do not recognize it, and moreover if we rely on it to meet guidelines when it isn't usable by end users, that opens us up to liability. I understand you do not have to take the laws of the United States into consideration, but we do. 

On a practical level there are two common approaches that we have seen used for accessible math being included in PDF 1.7. The first is to embed screen readable math in as invisible text "behind" elements, similar to what is done with Adobe's OCR layers. This method has the benefit of making the text also copy/pastable, but isn't easy to do manually in remediation. The other is to leverage the Alternate/Actual text features for tagged elements. Both need to be human readable, and it becomes labor intensive to implement, but this approach works and is actionable but end users. Both of these meet PDF/UA and WCAG standards, though I don't assume either of these approaches is especially automatable.

**Comment by Enivex** [comment]
@u-fischer @ErroneousBosch I'm not sure this is the right place for this seemingly LaTeX-specific conversation.

**Comment by jasonjgw** [comment]
I'm not convinced that implementing tagged PDF would be worth the effort for Typst developers. As a screen reader user, I wouldn't even know that a PDF file was tagged unless I happened to be using one of the few client-side implementations supporting it. So far as I know, the browser-based PDF readers in Chromium and Firefox don't support tags, and they're the default means of reading PDF documents encountered on the Web for users of those browsers.

By contrast, accessible HTML works everywhere, regardless of operating system and screen reader. Typst developers are already committed to implementing export to HTML. As long as this follows good accessibility practices, it will address the need so far as screen reader users are concerned. Obviously, this necessitates the use of MathML for mathematical content. We already have good braille and speech support for MathML (MathCAT and Speech Rule Engine are both reputable implementations). It's no surprise, for example, that the Arxiv project's accessibility strategy for LaTeX documents is to convert submitted papers to HTML format.

**Comment by domoritz** [comment]
Tagged PDFs (and other annotations like alt texts) do make a difference and are a requirement for example at the ASSETS conference: https://assets23.sigaccess.org/guidelines-policies.html. Also, screen readers will improve and use the tags when there are more tagged documents in the world.

I do believe it should be a priority for Typst to produce tagged documents (in addition to HTML). The issue with relying on HTML is that many conferences only accept PDFs and the PDF may often be the file type someone will encounter.

**Comment by ErroneousBosch** [comment]
@jasonjgw , thank you for this insight, and the information about Arxiv. We are trying to find a solution for our LaTeX users, and we have also been looking at Typst as an alternative. LaTeX HTML support through PanDoc either to MathML or MathJax output is what we found works best, and it would be nice to see a shift away from PDF.

**Comment by domoritz** [comment]
Typst is trending on HN and many comments mention support for HTML, accessibility, and tagged PDFs and consider it a blocker: https://news.ycombinator.com/item?id=41014941.

**Comment by mohe2015** [comment]
Please remember that this is a relatively new project with a small amount of developers that provide this for free.

**Comment by djturizo** [comment]
PDF accessibility is becoming is becoming compulsory at universities all over the worlds very quickly. LaTeX support for PDF tags is not at usable level yet. If the Typst devs manage to offer complete and automatic PDF tags quickly, this could become the inflexion point for Typst to supersede LaTeX at large.

**Comment by Enivex** [comment]
> PDF accessibility is becoming is becoming compulsory at universities all over the worlds very quickly.

I'll believe it when I see it. The tools just aren't there yet, neither for production nor consumption.

> LaTeX support for PDF tags is not at usable level yet.

Case in point

**Comment by samtalki** [comment]
Just want to concur that this is a giant opportunity for Typst; LaTeX has completely failed to meet this important challenge. I'm serving as an instructor this semester and tagging is completely non-functional in Beamer in particular, and I would switch to Typst instantly if I could more easily meet accessibility requirements.
@Enivex @djturizo

**Comment by domoritz** [comment]
Also hugely in favor of supporting accessible PDFs (in addition to HTML as an even more accessible format). Recently, Latex has been making a lot of progress on supporting tagged PDFs: https://latex-project.org/news/2024/07/08/tagging/.

**Comment by u-fischer** [comment]
@samtalki 

> LaTeX has completely failed to meet this important challenge.

It hasn't. We put lots of work and thought into this topic in the past years and as a result LaTeX was one of the first to be able to produce Well Tagged PDF (WTPDF) when the PDF association released that standard at the begin of the year and many documents can now be tagged automatically and easily with LaTeX.
This work will continue --- how fast depends also on founding, support from users e.g. to create test files and support from package and class authors who have to adapt their code.

>  I'm serving as an instructor this semester and tagging is completely non-functional in Beamer in particular. 

beamer is quite a challenge. The beamer class is quite large but stems from a time when only very few people thought about accessibility. Some discussion and also some minimal code to get beamer tagging aware is here https://github.com/latex3/tagging-project/issues/710.

**Comment by ErroneousBosch** [comment]
> I'll believe it when I see it. The tools just aren't there yet, neither for production nor consumption.

@Enivex I work for a major state university in the US in their college of engineering that does huge amounts of research. The university's policy is now that all materials, including PDFs from researchers, must meet [WCAG 2.0 AA compliance](https://www.w3.org/TR/WCAG20-TECHS/pdf). Tools or not, the requirement is in place and furthermore will be mandated by [US federal law starting April 2026](https://www.ada.gov/resources/2024-03-08-web-rule/). We have had to tell our researchers they can no longer use LaTeX without remediating each document. As I have explained in other discussions, ideals are fine to have but we are well into where the rubber has to meet the road.  If Typst can step up to create PDFs that meet the needed standards, it would be a fantastic solution. 

@u-fischer I'm curious to see what progress LaTeX has made and will have to revisit my evaluations

**Comment by laurmaedje** [comment]
Yes, HTML export helps (although it still is far from making it trivial). Work on tagged PDF is slowly progressing through that and https://github.com/typst/typst/pull/5420.

**Comment by Nicceboy** [comment]
Just to add to the list; in Finland, new global accessibility requirements will come into force, [starting from 28.6.2025](https://www.traficom.fi/en/traficom/accessibility/new-accessibility-requirements), as a result of the previously mentioned [European Accessibility Act](https://en.wikipedia.org/wiki/European_Accessibility_Act). From that day, the directive will be enforced across the EU, so I suppose everyone will need PDF/UA files pretty soon.

**Comment by Timtam** [comment]
Blind PhD student here, is there any update on this topic? Now that the EU accessibility act is active, I want to write my thesis and papers in a format that is as accessible as possible, and publishing will most likely happen as PDFs. I just learned about Typst and would like to replace my LateX workflow with it if possible, but creating accessible PDFs is definitely one of the must haves here.

**Comment by mohe2015** [comment]
To me it seems like there is very active progress in https://github.com/typst/typst/pull/6619

**Comment by Enivex** [comment]
> Blind PhD student here, is there any update on this topic? Now that the EU accessibility act is active, I want to write my thesis and papers in a format that is as accessible as possible, and publishing will most likely happen as PDFs. I just learned about Typst and would like to replace my LateX workflow with it if possible, but creating accessible PDFs is definitely one of the must haves here.

I believe the intention is that it should be a part of 1.14, which isn't too far away now. "Only" PDF/UA-1 though, not PDF/UA-2.

**Comment by Nicceboy** [comment]
> Blind PhD student here, is there any update on this topic? Now that the EU accessibility act is active, I want to write my thesis and papers in a format that is as accessible as possible, and publishing will most likely happen as PDFs. I just learned about Typst and would like to replace my LateX workflow with it if possible, but creating accessible PDFs is definitely one of the must haves here.

Fair warning that many journals/conferences still want LaTeX source and Typst is not always okay :(

**Comment by jasonjgw** [comment]
The basic HTML support already available in Typst 0.13.1 gave semantically correct output in my test, so I think this work is progressing in the right direction. When I was a "blind Ph.D. student" myself, I used LaTeX to typeset my thesis, and of course I've used it subsequently for published work. The potential advantage of Typst is that the extension packages will be able to customize the HTML output for the parts of the document that they render.

**Comment by saecki** [comment]
With #6619 merged I think this general issue has been addressed!
